File: src/ObjectManager/Object.Tes/FilePacks/Records/345-INGR.Ingredient.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class INGRRecord : Record, IHaveEDID, IHaveMODL
    {
        // TES3
        public struct IRDTField
        {
            public float Weight;
            public int Value;
            public int[] EffectId; // 0 or -1 means no effect
            public int[] SkillId; // only for Skill related effects, 0 or -1 otherwise
            public int[] AttributeId; // only for Attribute related effects, 0 or -1 otherwise

            public IRDTField(UnityBinaryReader r, int dataSize)
            {
                Weight = r.ReadLESingle();
                Value = r.ReadLEInt32();
                EffectId = new int[4];
                for (var i = 0; i < EffectId.Length; i++)
                    EffectId[i] = r.ReadLEInt32();
                SkillId = new int[4];
                for (var i = 0; i < SkillId.Length; i++)
                    SkillId[i] = r.ReadLEInt32();
                AttributeId = new int[4];
                for (var i = 0; i < AttributeId.Length; i++)
                    AttributeId[i] = r.ReadLEInt32();
            }
        }

        // TES4
        public class DATAField
        {
            public float Weight;
            public int Value;
            public uint Flags;

            public DATAField(UnityBinaryReader r, int dataSize)
            {
                Weight = r.ReadLESingle();
            }

            public void ENITField(UnityBinaryReader r, int dataSize)
            {
                Value = r.ReadLEInt32();
                Flags = r.ReadLEUInt32();
            }
        }

        public override string ToString() => $"INGR: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model Name
        public STRVField FULL; // Item Name
        public IRDTField IRDT; // Ingrediant Data //: TES3
        public DATAField DATA; // Ingrediant Data //: TES4
        public FILEField ICON; // Inventory Icon
        public FMIDField<SCPTRecord> SCRI; // Script Name
        // TES4
        public List<ENCHRecord.EFITField> EFITs = new List<ENCHRecord.EFITField>(); // Effect Data
        public List<ENCHRecord.SCITField> SCITs = new List<ENCHRecord.SCITField>(); // Script effect data

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL": if (SCITs.Count == 0) FULL = new STRVField(r, dataSize); else ArrayUtils.Last(SCITs).FULLField(r, dataSize); return true;
                case "FNAM": FULL = new STRVField(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize); return true;
                case "IRDT": IRDT = new IRDTField(r, dataSize); return true;
                case "ICON":
                case "ITEX": ICON = new FILEField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                    //
                case "ENIT": DATA.ENITField(r, dataSize); return true;
                case "EFID": r.SkipBytes(dataSize); return true;
                case "EFIT": EFITs.Add(new ENCHRecord.EFITField(r, dataSize, format)); return true;
                case "SCIT": SCITs.Add(new ENCHRecord.SCITField(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-INGR.Ingredient.swift
//
//  INGRRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class INGRRecord: Record, IHaveEDID, IHaveMODL {
    // TES3
    public struct IRDTField {
        public let weight: Float
        public let value: Int32
        public var effectId: [Int32] // 0 or -1 means no effect
        public var skillId: [Int32] // only for Skill related effects, 0 or -1 otherwise
        public var attributeId: [Int32] // only for Attribute related effects, 0 or -1 otherwise

        init(_ r: BinaryReader, _ dataSize: Int) {
            weight = r.readLESingle();
            value = r.readLEInt32();
            effectId = [Int32](); effectId.reserveCapacity(4)
            for _ in 0..<4 { effectId.append(r.readLEInt32()) }
            skillId = [Int32](); skillId.reserveCapacity(4)
            for _ in 0..<4 { skillId.append(r.readLEInt32()) }
            attributeId = [Int32](); attributeId.reserveCapacity(4)
            for _ in 0..<4 { attributeId.append(r.readLEInt32()) }
        }
    }

    // TES4
    public class DATAField {
        public let weight: Float
        public var value: Int32 = 0
        public var flags: UInt32 = 0

        init(_ r: BinaryReader, _ dataSize: Int) {
            weight = r.readLESingle()
        }

        func ENITField(_ r: BinaryReader, _ dataSize: Int) {
            value = r.readLEInt32()
            flags = r.readLEUInt32()
        }
    }

    public override var description: String { return "INGR: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var MODL: MODLGroup? = nil // Model Name
    public var FULL: STRVField! // Item Name
    public var IRDT: IRDTField! // Ingrediant Data //: TES3
    public var DATA: DATAField! // Ingrediant Data //: TES4
    public var ICON: FILEField! // Inventory Icon
    public var SCRI: FMIDField<SCPTRecord>! // Script Name
    // TES4
    public var EFITs = [ENCHRecord.EFITField]() // Effect Data
    public var SCITs = [ENCHRecord.SCITField]() // Script effect data

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "FULL": if SCITs.count == 0 { FULL = r.readSTRV(dataSize) } else { SCITs.last!.FULLField(r, dataSize) }
        case "FNAM": FULL = r.readSTRV(dataSize)
        case "DATA": DATA = DATAField(r, dataSize)
        case "IRDT": IRDT = IRDTField(r, dataSize)
        case "ICON",
             "ITEX": ICON = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
            //
        case "ENIT": DATA.ENITField(r, dataSize)
        case "EFID": r.skipBytes(dataSize)
        case "EFIT": EFITs.append(ENCHRecord.EFITField(r, dataSize, format))
        case "SCIT": SCITs.append(ENCHRecord.SCITField(r, dataSize))
        default: return false
        }
        return true
    }
}
